File: answers/other/coding/Sum Root to Leaf Numbers.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        
        return self.helper(root, 0)

    def helper(self, root, path):
    	if root:
    		path = path * 10 + root.val
    		if None == root.left and None == root.right:
    			return path
    		else:
    			return self.helper(root.left, path) + self.helper(root.right, path)
    	else:
    		return 0

File: answers/leetcode/Sum Root to Leaf Numbers/Sum Root to Leaf Numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int sumNumbers(TreeNode root) {
        return this.dfs(root, 0);
    }

    private int dfs(TreeNode root, int num) {
        if (null == root) {
            return 0;
        } else if (null == root.left && null == root.right) {
            return num * 10 + root.val;
        } else {
            return this.dfs(root.left, num * 10 + root.val) + this.dfs(root.right, num * 10 + root.val);
        }
    }
}